### PR TITLE
Add fusion support (resolves #1)

### LINKED ISSUE #1: Add fusion support

### PR DESCRIPTION
**Pull request by jpfeil**
resolves #1

### CODE REVIEW

File: fusion.py [hunk 1]
```diff
@@ -0,0 +1,519 @@
...
+                    acceptor_in_frame_adjustment = scan_frame(acceptor_bounds.rstart)
+                    if acceptor_in_frame_adjustment is None:
+                        continue
+
+                # We can still generate a fusion transcript without a reference for the acceptor gene
+                except KeyError:
+                    pass
+
+                # Figure out if frame-shift occurred as a result of the fusion
+                acceptor_frame_shift = False
```
Comment (arkal) [inline fusion.py]: What is this used for?

File: fusion.py [hunk 2]
```diff
@@ -0,0 +1,519 @@
...
+                    acceptor_frame_shift = True
+
+                # Use the donor and acceptor sequences to generate the fusion transcript
+                fusion_seq = in_frame_donor_seq + junctionSeq2
+                logging.debug(fusion_seq)
+
+                print("JUNCTIONSEQ2")
+                print(junctionSeq2)
+
+                # The fusion peptide for MHC binding prediction includes all sliding windows that would generate a
```
Comment (arkal) [inline fusion.py]: 100 chars

File: fusion.py [hunk 3]
```diff
@@ -0,0 +1,519 @@
...
+                # still output the fusion peptide because there is at least one fusion sequence that can be generated
+                # from it.
+                if len(fusion_peptide) != 2 * peplen - 2:
+                    logging.warning('Fusion peptide %s-%s is malformed.\n'
+                                    'Expected peptide of length %d, got %d' % (donor_name,
+                                                                               acceptor_name,
+                                                                               2 * peplen - 2,
+                                                                               len(fusion_peptide)))
+
+                # print(call.hugo1, call.hugo2)
```
Comment (arkal) [inline fusion.py]: Remove debug lines

File: test/test_input/fusion_results.bedpe [hunk 4]
```diff
@@ -0,0 +1,11 @@
+# chr1	start1	end1	chr2	start2	end2	name	score	strand1	strand2	junctionSeq1	junctionSeq2	hugo1	hugo2
```
Comment (arkal) [inline test/test_input/fusion_results.bedpe]: Is this file used anywhere?

File: test/test_input/hook3_ret.results.tsv [hunk 5]
```diff
@@ -0,0 +1 @@
+8	.	42823357	10	43612032	.	ENSG00000168172-ENSG00000165731	.	+	+	TAAGAAAGGCCAACGCAGCGCGAAGTCAACTTGAAACCTACAAGAGACAG	GAGGATCCAAAGTGGGAATTCCCTCGGAAGAACTTGGTTCTTGGAAAAAC	HOOK3	RET
```
Comment (arkal) [inline test/test_input/hook3_ret.results.tsv]: Is this file used anywhere?

File: test/test_transgene.py [hunk 6]
```diff
@@ -65,9 +65,15 @@ def _test_transgene(self, use_RNA=None):
         params.no_json_dumps = False
         params.log_level = 'DEBUG'
         params.reject_threshold = 5
-        params.rna_file  = self._get_input_path('test_input/test.bam') if use_RNA else None
-        params.input_file = open(self._get_input_path('test_input/test.pc_translations.fa'))
+        params.rna_file = self._get_input_path('test_input/test.bam') if use_RNA else None
+        params.peptide_file = open(self._get_input_path('test_input/test.pc_translations.fa'))
+        params.peptide_file = None
```
Comment (arkal) [inline test/test_transgene.py]: ?

File: test/test_transgene.py [hunk 7]
```diff
@@ -65,9 +65,15 @@ def _test_transgene(self, use_RNA=None):
...
         params.snpeff_file = open(self._get_input_path('test_input/snpeff_test.vcf'))
+        params.snpeff_file = None
```
Comment (arkal) [inline test/test_transgene.py]: ?

File: transgene.py [hunk 8]
```diff
@@ -433,7 +389,8 @@ def insert_snvs(chroms, snvs, tumfile, normfile, peplen, rna_bam=None):
     :param dict chroms: Contains the peptides in the form of dicts where keys hold the protein name
     and the values holds the sequence.
     :param dict snvs: Contains the snvs parsed from teh input SnpEFF file.
-    :param file outfile: The file to write output to.
+    :param file tumfile: The file to write tumor output to.
```
Comment (arkal) [inline transgene.py]: whoops i overlooked that. Thanks.

File: transgene.py [hunk 9]
```diff
@@ -584,31 +547,18 @@ def parse_peptides(tumfile, normfile, prefix, suffix):
...
     """
-    This tool accepts a vcf of translated mutations -- i.e. a file that has information of the
-    mutation in genomic space (e.g. chr1:47403668) and in proteomic space (e.g.
-    ENST00000371904.4:D113N) -- and outputs files that contain n-mer peptides (n>0) that can be
-    passed to MHC:peptide binding prediction tools.
+    Loads GENCODE translated transcript sequences into a dictionary
 
-    Transgene can be run in an RNA-aware mode where it will only accept mutations that have evidence
-    in an RNA-seq bam (Mapped to the genome reference).
+    :param infile: GENCODE file object
```
Comment (arkal) [inline transgene.py]: type

File: transgene.py [hunk 10]
```diff
@@ -584,31 +547,18 @@ def parse_peptides(tumfile, normfile, prefix, suffix):
...
+    :return: Mapping of concatenated Ensembl gene IDs, Ensembl transcript IDs, and HUGO names to protein sequences
+    :rtype; dict
```
Comment (arkal) [inline transgene.py]: :rtype:

File: transgene.py [hunk 11]
```diff
@@ -548,11 +505,15 @@ def parse_peptides(tumfile, normfile, prefix, suffix):
             hugo_gene = pep_name[2]
             transcript_mutation = '_'.join([pep_name[1]]+pep_name[3:])
             if peptides[(gene_name, hugo_gene)] == 0:
-                peptides[(gene_name, hugo_gene)] = {transcript_mutation:
-                                                        (pep_seq, normal_peptides[full_pep_name])}
+                try:
+                    peptides[(gene_name, hugo_gene)] = {transcript_mutation: (pep_seq, normal_peptides[full_pep_name])}
```
Comment (arkal) [inline transgene.py]: 100 chars

File: fusion.py [hunk 12]
```diff
@@ -0,0 +1,519 @@
...
+
+def get_transcriptome_data(infile):
+    """
+    Parses Gencode transcript FASTA file and returns CDS sequences keyed by the transcript ID
+
+    :param file infile: Gencode FASTA file object
+    :return: Dictionary of transcripts keyed by transcript ID and dictionary of transcript IDs keyed by gene ID
+    :rtype: tuple
+    """
+    regex = r"(?P<transcript_id>ENST[0-9A-Z]+.\d+)\|(?P<gene_id>ENSG[0-9A-Z]+.\d+).*CDS:(?P<start>\d+)-(?P<stop>\d+)"
```
Comment (jpfeil) [inline fusion.py]: Yes

File: fusion.py [hunk 13]
```diff
@@ -0,0 +1,519 @@
...
+        elif chr == 'D':
+            deletions += num
+
+    # Next grab the alignment statistics
+    string = StringIO()
+    alignment.dump(out=string)
+    dump = string.getvalue()
+    string.close()
+
+    # If it's not a near perfect match, then the quality of the assembly may not be good
```
Comment (jpfeil) [inline fusion.py]: SNVs are already accounted for in the de novo assembled transcript.

File: fusion.py (hunk 1 above)
Comment (jpfeil) [inline fusion.py]: I'm going to output peptides that result from the frameshift.

File: test/test_input/fusion_results.bedpe (hunk 4 above)
Comment (jpfeil) [inline test/test_input/fusion_results.bedpe]: This file is used during the test.

File: test/test_input/hook3_ret.results.tsv (hunk 5 above)
Comment (jpfeil) [inline test/test_input/hook3_ret.results.tsv]: This is an older file that is no longer used during the test. I will remove it.

File: test/test_transgene.py (hunk 6 above)
Comment (jpfeil) [inline test/test_transgene.py]: This was for testing the fusion module, but I need to remove it now.

File: fusion.py (hunk 1 above)
Comment (arkal) [inline fusion.py]: Is that logic in right now or is this a future addition? It didn't look like that variable was used anywhere but i could be wrong.

File: fusion.py (hunk 13 above)
Comment (arkal) [inline fusion.py]: 👍

**Comment by arkal** [comment]
What is the progress on this?

**Comment by jpfeil** [comment]
@arkal Thanks for checking in! I'm ready for another review.

**Comment by jpfeil** [comment]
@arkal -- Ready for review

File: common.py [hunk 14]
```diff
@@ -0,0 +1,130 @@
+import re
+
+
+def read_fasta(input_file, alphabet):
+    """
+    This module will accept an input fasta and yield individual sequences
+    one at a time.
```
Comment (arkal) [inline common.py]: resolved

File: fusion.py [hunk 15]
```diff
@@ -0,0 +1,519 @@
...
+base2 = 'TTTTCCCCAAAAGGGGTTTTCCCCAAAAGGGGTTTTCCCCAAAAGGGGTTTTCCCCAAAAGGGG'
+base3 = 'TCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAGTCAG'
+genetic_code = dict((''.join([b1, b2, b3]), aa) for aa, b1, b2, b3 in zip(amino, base1, base2, base3))
+
+
+def get_transcriptome_data(infile):
+    """
+    Parses Gencode transcript FASTA file and returns CDS sequences keyed by the transcript ID
+
+    :param file infile: Gencode FASTA file object
```
Comment (arkal) [inline fusion.py]: resolved

File: fusion.py [hunk 16]
```diff
@@ -0,0 +1,519 @@
...
+            # GTF is one-based
+            start = int(match.group("start")) - 1
+            stop = int(match.group("stop"))
+            cds = seq[start: stop]
+
+            gene_transcripts[gene_id].append(transcript_id)
+            transcript_cds[transcript_id] = cds
+    return transcript_cds, gene_transcripts
+
+
```
Comment (arkal) [inline fusion.py]: resolved

File: fusion.py [hunk 17]
```diff
@@ -0,0 +1,519 @@
...
+    """
+    Generates fusion peptides for MHC binding prediction. If the junction sequence is not
+    provided for the donor and acceptor, a theoretical sequence is inferred using a
+    genome annotation file.
+
+    :param dict transcriptome: Dictionary of transcripts keyed by transcript ID
+    :param list[BEDPE] fusion_calls: List of BEDPE namedtuples
+    :param dict[list] gene_transcripts: Dictionary of transcript IDs keyed by gene ID
+    :param int peplen: Length of n-mer peptide
+    :param file outfile: File object to write n-mer peptides in FASTA format
```
Comment (arkal) [inline fusion.py]: resolved

File: fusion.py (hunk 1 above)
Comment (arkal) [inline fusion.py]: resolved

**Comment by jpfeil** [comment]
@arkal -- I squashed and added a commit message.

**Pull request merged**